scripts/GWA_TBX/PG04/PG04_WaterQualityWorkflow_SNAP_l3binning.py

```diff
@@ -82,13 +82,13 @@ def create_graph(tempdir, files, rows, new_filename, sensor):
             text_file.write('              <outputCounts>false</outputCounts>\n')
             text_file.write('              <outputSums>false</outputSums>\n')
             text_file.write('          </aggregator>\n')
-            text_file.write('          <aggregator>\n')
-            text_file.write('              <type>AVG</type>\n')
-            text_file.write('              <varName>chl</varName>\n')
-            text_file.write('              <weightCoeff>0.0</weightCoeff>\n')
-            text_file.write('              <outputCounts>false</outputCounts>\n')
-            text_file.write('              <outputSums>false</outputSums>\n')
-            text_file.write('          </aggregator>\n')
+#            text_file.write('          <aggregator>\n')
```

Comment:
You should really add an XML file to the repo that you read, modify, and write - if it is modified at all? It does not seem so.

Reply:
We wanted to keep the "graph" (that's what this xml is called in context of gpt) inside the python code. Therefore it is independent from any auxiliary data. Writing the xml file take only a split second and thus does not affect the processing at all. After processing the file and the other generated files are removed from the users temp folder.

Reply:
> Therefore it is independent from any auxiliary data.

I'd say a python file containing lines of data written unmodified and one-by-one to a text file is as auxiliary as a text file in the same folder.

But if you want to keep it in the code, you could put it in a block quote that preserves the newline characters:
```python
gpt_str = """My giant
GPT graph
with many
lines"""
with open('out.xml', 'w') as f:
    f.write(gpt_str)
```

Reply:
Is there a (technical) reason why the gpt operators are typed directly into this script instead of being added to the processing_gpf plugin and than used in a graph? The latter way would provide much more benefit to the broader QGIS-SNAP community (outside of GWA) and would improve reuseability and maintainability of the implementation.

Reply:
@j08lue: We can do that if needed. But if it's ok, I will keep it as it is now to save us some extra effort.

@radosuav: A graph can be compared to a QGIS model. If you add these additional processors the processing_gpf plugin it's a nice thing for the users, but I will not change anything to my current workflow. The SNAP operators cannot be combined in the normal model builder, only the dedicated GPF Graph Builder. But this tool only does what I have done, it builds an XML graph. A question back to you: What do I have to do to create new GUIs for SNAP operators in Processing Tools?

Reply:
It's true that it won't change anything for your current workflow. But it will make it easier if you want to use the same operators in the future and will make them accessible to a much wider user base. 

SNAP operators can be used directly in workflows and QGIS models as well as be combined in the GPF Graph Builder. In normal modeler SNAP algorithms can be combined with any other Processing algorithm but the tradeoff is efficiency because a separate GPF graph is created for each SNAP operator. In GPF Graph builder only SNAP algorithms can be used but they are all combined into one big graph (like your script) which reduces the I/O operations. So if you want to use multiple SNAP operators consecutively then the "optimal" solution is to create a graph using GPF graph builder and then use it in the normal modeler or workflow.

To add a new SNAP operator to Processing all that is needed is a text file describing the input and output parameters. The text file needs to be placed in one of the description file directories (for [generic SNAP algorithms](https://github.com/DHI-GRAS/processing_gpf/tree/GW-A_dev/snap_generic_description), [S1TBX](https://github.com/DHI-GRAS/processing_gpf/tree/GW-A_dev/s1tbx_description), [S2TBX](https://github.com/DHI-GRAS/processing_gpf/tree/GW-A_dev/s2tbx_description) or [S3TBX](https://github.com/DHI-GRAS/processing_gpf/tree/GW-A_dev/s3tbx_description)). For more details have a look [here](https://github.com/DHI-GRAS/processing_gpf/issues/41) and [here](https://github.com/DHI-GRAS/processing_gpf/issues/43).

Cheers!

Reply:
The second paragraph was already understood by me. I was just thinking about our own application. Anyhow, I thinks there is nothing more to add.

Your last paragraph is most interesting to me, as I already tried to understand how it works and was wondering if all is needed are these small text file. As this seems to be the case, we can have a look into writing some additional ones for other operators.

Cheers,
Jan

Reply:
I already merged this PR, but I'd be very happy to see some improvements in another PR, @JanWev. We will have to maintain this over the next years, so when SNAP changes its algorithm names or something breaks for other reasons, it would be nice to have a clean place to adjust your graph.

Moving the graph into `processing_gpf` would be premium, but either an auxiliary file or a code block inside the python file would be okay, too.